Added documentation for Spotlight HID++ (#150)
jahnf/Projecteur: doc/LogitechSpotlightHID++.md

| DEVICE_GROUPS            | 0x0006     |
| DEVICE_FRIENDLY_NAME     | 0x0007     |
| KEEP_ALIVE               | 0x0008     |
| RESET                    | 0x0020     |
| CRYPTO_ID                | 0x0021     |
| TARGET_SOFTWARE          | 0x0030     |
| WIRELESS_SIGNAL_STRENGTH | 0x0080     |
| DFUCONTROL_LEGACY        | 0x00C0     |
| DFUCONTROL_UNSIGNED      | 0x00C1     |
| DFUCONTROL_SIGNED        | 0x00C2     |
| DFU                      | 0x00D0     |

A more extensive list of known feature codes are documented by [Solaar project](https://github.com/pwr-Solaar/Solaar/blob/master/docs/features.md). Some of the feature code relevant for Logitech Spotlight is defined in [hidpp.h](../src/hidpp.h).

```c++
enum class FeatureCode : uint16_t {
  Root                 = 0x0000,
  FeatureSet           = 0x0001,
  FirmwareVersion      = 0x0003,
  DeviceName           = 0x0005,
  Reset                = 0x0020,
  DFUControlSigned     = 0x00c2,
  BatteryStatus        = 0x1000,
  PresenterControl     = 0x1a00,
  Sensor3D             = 0x1a01,
  ReprogramControlsV4  = 0x1b04,
  WirelessDeviceStatus = 0x1db4,
  SwapCancelButton     = 0x2005,
  PointerSpeed         = 0x2205,
};
```

> No single Logitech device support all feature codes, rather they support a limited range of features and corresponding feature codes. Inside the device the supported feature codes are mapped to a Index (or Feature Index). This mapping is called FeatureSet table. For any device, the Root Feature Code (0x0000) has index of 0x00. Root Feature Index (0x00) is used to d

Reviewer: ```suggestion
No single Logitech device supports all feature codes, rather they support a limited range of features and corresponding feature codes. Inside the device the supported feature codes are mapped to an index (or Feature Index). This mapping is called FeatureSet table. For any device, the Root Feature Code (`0x0000`) has an index of `0x00`. Root Feature Index (`0x00`) is used to d
```
Last sentence incomplete.

> For any device, the feature index corresponding to any feature code can be obtained by using Root Feature Index (0x00) by sending signal

Reviewer: ```suggestion
For any device, the feature index corresponding to any feature code can be obtained by using Root Feature Index (0x00) by sending the request message
```

{0x10, 0x01, 0x00, 0x0d, Feature_Code(2 bytes), 0x00} (here, function code is 0x00 and software identification code is 0x0d in forth byte). If the return message is not an error message, fifth byte in response is Feature index and 6th byte is Feature type (See below).

> Software can get the entire FeatureSet table for the device with following algorithm:

Reviewer: ```suggestion
The application can retrieve the entire FeatureSet table for the device with following steps:
```

  1. Get the number of features supported by device:

> 		a. Get Feature Index corresponding to Feature Set code (0x0001).

Reviewer: ```suggestion
		a. Get the _Feature Index_ corresponding to the Feature Set code (`0x0001`).
```

> 		b. If the Feature Index obtained is 0x01 then get the count by sending {0x10, 0x01, 0x01, 0x0d, 0x00, 0x00, 0x00} (3rd byte is Feature Index; function code is 0x00 and software identification code is 0x0d in forth byte). In the response, 5th byte will be the number of features supported except root feature. As stated above, Root feature always has Feature Index 0x00. Hence, total number of features supported is 1 + count obtained in response.

Reviewer: ```suggestion
		b. If the Feature Index obtained is `0x01` then request the feature count by sending `{0x10, 0x01, 0x01, 0x0d, 0x00, 0x00, 0x00}` (3rd byte is Feature Index; function code is `0x00` and software identification code is `0x0d` in forth byte). In the response, the 5th byte will be the number of features supported, except the root feature. As stated above, the Root feature always has the Feature Index `0x00`. Hence, total number of features supported is one plus the count obtained in the response.
```

>   2. Iterate over Feature Index 1 to (number of feature supported) and send the signal (assuming Feature_Index for feature set is 0x01; third byte) {0x10, 0x01, 0x01, 0x1d, Feature_Index, 0x00, 0x00} (function code is 0x10 and software identification code is 0x0d in forth byte). The response will have HID++ Feature Code in (5 and 6 byte) as uint16_t and Feature Type in 7th byte for valid Feature Index. In Feature Type byte, if 7 and 8 bit are ON then it means Software hidden and Obsolete feature respectively. So, Software_Hidden = (Feature_Type & (1<<6)) and Obsolete_Feature = (Feature_Type & (1<<7)). Software Hidden or Obsolete feature should not be handled by any software. In case Feature Index is not valid (i.e.,feature index > number of feature supported) then 0x0200 will be in (5 and 6 byte).

Reviewer: ```suggestion
  2. Iterate over the Feature Indexes 1 to  the number of features supported and send the request (assuming Feature_Index for feature set is `0x01`; third byte) `{0x10, 0x01, 0x01, 0x1d, Feature_Index, 0x00, 0x00}` (function code is `0x10` and software identification code is `0x0d` in forth byte). The response will contain the HID++ Feature Code at byte 5 and 6 as `uint16_t` and the Feature Type at byte 7 for a valid Feature Index. In the Feature Type byte, if bits 7 is set this means _Software Hidden_, if bit 8 is set this means _Obsolete feature_. So, Software_Hidden = (Feature_Type & (1<<6)) and Obsolete_Feature = (Feature_Type & (1<<7)). Software Hidden or Obsolete features should not be handled by any application. In case the Feature Index is not valid (i.e.,feature index > number of feature supported) then `0x0200` will be in the response at byte 5 and 6.
```

> The FeatureSet table may change with firmware update. Hence, software should cache FeatureSet table along with Firware version and only read FeatureSet table again if firmware version has changed. This logic for getting FeatureSet table from device is coded in `populateFeatureTable` method in `FeatureSet` class in [hidpp.h](../src/hidpp.h).

Reviewer: ```suggestion
The FeatureSet table for a device may change with a firmware update. The software should cache FeatureSet table along with Firware version and only read FeatureSet table again if the firmware version has changed. This logic for getting FeatureSet table from device is coded in `populateFeatureTable` method in `FeatureSet` class in [hidpp.h](../src/hidpp.h).
```

# Resetting Logitech Spotlight device

> Depending on connection mode (USB dongle or Bluetooth), the logitech spotlight device can be reset with following HID++ message from software:

Reviewer: ```suggestion
Depending on the connection mode (USB dongle or Bluetooth), the Logitech Spotlight device can be reset with following HID++ message from the application:
```

>   1. Reset USB dongle by sending following commands in sequence

Reviewer: ```suggestion
  1. Reset the USB dongle by sending following commands in sequence
```

```
	{0x10, 0xff, 0x81, 0x00, 0x00, 0x00, 0x00} //get wireless notification and software connection status
	{0x10, 0xff, 0x80, 0x00, 0x00, 0x01, 0x00} // set sofware bit to false
	{0x10, 0xff, 0x80, 0x02, 0x02, 0x00, 0x00} // initialize the USB dongle
	{0x10, 0xff, 0x80, 0x00, 0x00, 0x09, 0x00} // set sofware bit to true
```

>   2. Load the FeatureSet table for device (from pre-existing cache or from device if firmware version has changed by calling `populateFeatureTable` method in `FeatureSet` class in [hidpp.h](../src/hidpp.h)).

Reviewer: ```suggestion
  2. Load the FeatureSet table for the device (from pre-existing cache or from the device if firmware version has changed by calling `populateFeatureTable` method in `FeatureSet` class in [hidpp.h](../src/hidpp.h)).
```

>   3. Reset the Spotlight device with Feature index for Reset Feature Code from FeatureSet table. If the feature index for reset feature code is 0x05, then HID++ message will be {0x10, 0x01, 0x05, 0x1d, 0x00, 0x00, 0x00}.

Reviewer: ```suggestion
  3. Reset the Spotlight device with the Feature index for the Reset Feature Code from the FeatureSet table. If the feature index for reset feature code is `0x05`, then HID++ message will be `{0x10, 0x01, 0x05, 0x1d, 0x00, 0x00, 0x00}`.
```

> In addition to these steps, the Projecteur also pings the device by sending {0x10, 0x01, 0x00, 0x1d, 0x00, 0x00, 0x5d} (function code 0x10 and software identification code 0x0d in forth byte). The response to this ping contain the HID++ version supported by the device.

Reviewer: ```suggestion
In addition to these steps, the Projecteur also pings the device by sending `{0x10, 0x01, 0x00, 0x1d, 0x00, 0x00, 0x5d}` (function code `0x10` and software identification code `0x0d` in forth byte). The response to this ping contains the HID++ version supported by the device.
```

> The Projecteur further configure the Logitech device to send `Next Hold` and `Back Hold` events and reset the pointer speed to normal with following HID++ commands.

Reviewer: ```suggestion
Further, Projecteur configures the Logitech device to send `Next Hold` and `Back Hold` events and resets the pointer speed to a default value with following HID++ commands:
```

```
// enable next button hold (0x07 - Feature Index for ReprogramControlsV4 Feature Code, 0xda - next button, 0x33 - hold event)
{0x11, 0x01, 0x07, 0x3d, 0x00, 0xda, 0x33, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
// back button hold (0x07 - Feature Index for ReprogramControlsV4 Feature Code, 0xdc - back button, 0x33 - hold event)
{0x11, 0x01, 0x07, 0x3d, 0x00, 0xdc, 0x33, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
// Reset pointer Speed (0x0a - Feature Index for Reset Feature Code, 0x14 - 5th level pointer speeed (can be between 0x10-0x19))
{0x10, 0x01, 0x0a, 0x1d, 0x14, 0x00, 0x00}
```

> These initialization steps are implemented in `initialize` method of `SubHidppConnection` class in [device.h](../src/device.h). it should be noted that after the reprogramming of Next and Back button in first and second message, the spotlight device send relative mouse move data through HID++ messages (not regular input events to OS). These events are handled in `onHidppDataAvailable` method in `Spotlight` class in [spotlight.h](../src/spotlight.h).

Reviewer: ```suggestion
These initialization steps are implemented in `initialize` method of `SubHidppConnection` class in [device.h](../src/device.h). After reprogramming the Next and Back buttons, the first and second response of the spotlight device contain relative mouse move data through HID++ messages (not regular input events to the OS). These events are handled in `onHidppDataAvailable` method in `Spotlight` class in [spotlight.h](../src/spotlight.h).
```

> For the sake of completeness, it should be noted that the official Logitech Spotlight software reprogram the click and double click events too by following HID++ commands:

Reviewer: ```suggestion
For completeness, it should be noted that the official Logitech Spotlight software reprogram the click and double click events too by following HID++ commands:
```

```
// Send click event as HID++ message (0x07 - Feature Index for ReprogramControlsV4 Feature Code, 0xd8-click button, 0x33- hold event)
{0x11, 0x01, 0x07, 0x3d, 0x00, 0xd8, 0x33, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
// Send double click as HID++ message (0x07 - Feature Index for ReprogramControlsV4 Feature Code, 0xdf - double click)
{0x11, 0x01, 0x07, 0x3d, 0x00, 0xdf, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
```

> However, Projecteur does not send these packages. Instead it grab the events from mouse event device associated with Spotlight. The approach taken by Projecteur is advantageous as it help in implementing Input Mapping feature that official Logitech sofware lacks. However, this approach also make porting Projecteur to different platform somewhat difficult.

Reviewer: ```suggestion
Projecteur does not send these packages. Instead it grab the events from mouse event device associated with Spotlight. The approach taken by Projecteur is advantageous as it help in implementing Input Mapping feature that official Logitech sofware lacks. However, this approach also makes porting Projecteur to different platforms more difficult.
```

# Wireless Notification on Activation/Deactivation of Spotlight Device

> The Spotlight device send wireless notification if it get activated. Wireless notification will be short HID++ message if Spotlight device is connected through USB dongle. Otherwise, it will be a long HID++ message.

Reviewer: ```suggestion
The Spotlight device sends a wireless notification if it gets activated. Wireless notification will be short HID++ message if the Spotlight device is connected through USB. Otherwise, it will be a long HID++ message.
```

> For short HID++ wireless notification, the third byte will be 0x41. In such message, 6th bit in 5th byte show activation status of spotlight device. If the 6th bit is 0 then device got active otherwise the device got deactivated.

Reviewer: ```suggestion
For short HID++ wireless notifications, the third byte will be `0x41`. In this message, the 6th bit in 5th byte shows the activation status of spotlight device. If the 6th bit is 0 then device got active, otherwise the device got deactivated.
```

> Long HID++ wireless notification is only received for device activation. In such message, third byte will be Feature Index for Wireless Notification Feature Code.

Reviewer: ```suggestion
A long HID++ wireless notification is only received for device activation. In this message, third byte will be the Feature Index for the Wireless Notification Feature Code.
```

# Vibration support

> The spotlight device can vibrate if HID++ message {0x10, 0x01, (Feature Index for Presenter Control Feature Code), 0x1d, length, 0xe8, intensity} is sent to it. In the message, length can range between 0x00 to 0x0a.

Reviewer: ```suggestion
The spotlight device can vibrate if the HID++ message `{0x10, 0x01, (Feature Index for Presenter Control Feature Code), 0x1d, length, 0xe8, intensity}` is sent to it. In the message, length can range between `0x00` to `0x0a`.
```

# Note

> All of HID++ commands listed above result in response from Spotlight device. Many a time these response are just the acknowledgement of the HID++ commands sent from PC. However, in some of response Spotlight device can send useful information. To check processing of these responses please check the `onHidppDataAvailable` method in `Spotlight` class in [spotlight.h](../src/spotlight.h).

Reviewer: ```suggestion
All of the HID++ commands listed above result in response messages from the Spotlight device. For most messages these responses are just the acknowledgements of the HID++ commands sent by the application. However, some responses  from the Spotlight device contain useful information. For details on processing responses see the  `onHidppDataAvailable` method in the `Spotlight` class in [spotlight.h](../src/spotlight.h).
```

> For more information about HID++ protocol in general please check files uploaded to [this folder](https://drive.google.com/drive/folders/0BxbRzx7vEV7eWmgwazJ3NUFfQ28?resourcekey=0-dQ-Lx1FORQl0KAdOHQaE1A) or checkout [logitech-hidpp module](https://github.com/torvalds/linux/blob/master/drivers/hid/hid-logitech-hidpp.c) code in linux kernel source.

Reviewer: I don't like linking to some google drive link for legal reasons
1. I don't know what documents are on there 
2. Event if the documents that are on there are totally fine right now, I don't know what happens in the future and I can't monitor that drive share.
